Handle user interactions

elayabharath opened the pull request:
+ Handle analysis tabs switch, toggle comparison data, toggle filter options, change filter options
+ Refactored the filter & chart methods to computeXYZAndRender()
+ Common UI IDs for dropdowns, checkboxes (named it as utils - please let me know if may be I can change to selectors.dart)

elayabharath committed:
Handle UI interactions

marianamarasoiu commented on lib/controller.dart:
```diff
@@ -349,14 +362,23 @@ void command(UIAction action, Data data) {
+      _activeFilters = {};
+      _filterValues = {};
+      _comparisonFilterValues = {};
+      view.removeFiltersWrapper();
+      view.removeAllChartWrappers();
+      _computeFilterDropdownsAndRender();
+      _computeChartBucketsAndRender();
```
what's the performance of clearing up the UI and reloading everything? is there any flickering? If yes, we might need to look at incrementally updating the UI...

elayabharath replied:
At the moment I don't see flickers as such. 

But if I have to build this in..
1. The charts need to be taken out as they are highly likely to be different values for each of the tabs. Also the chartjs doesn't refresh the "on-air" image symbols without throwing away the canvas and re-rendering, say when you want to toggle on / off comparison.
2. For the filter dropdowns, I was not able to maintain the order of filter keys as I could only append them in the end after comparing the diff of filter keys across both tabs. Just realised as I was writing this I could use hidden parameter against each of the filter rows. - Will update this PR with this functionality soon.

marianamarasoiu commented on lib/view.dart:
```diff
@@ -107,6 +114,27 @@ void clearContentTab() {
+void removeFiltersWrapper() {
+  if (filtersWrapper == null) {
+    logger
+        .error('Trying to remove non existant selector #${FILTERS_WRAPPER_ID}');
```
nit:
```suggestion
        .error('Trying to remove non-existent selector #${FILTERS_WRAPPER_ID}');
```

elayabharath replied:
Fixed

marianamarasoiu commented on lib/view.dart:
```diff
@@ -107,6 +114,27 @@ void clearContentTab() {
+  for (var wrapper in chartWrappers) {
+    if (wrapper == null) {
+      logger.error(
+          'Trying to remove non existant selector .${CHART_WRAPPER_CLASSNAME}');
```
nit:
```suggestion
          'Trying to remove non-existent selector .${CHART_WRAPPER_CLASSNAME}');
```

elayabharath replied:
Fixed

marianamarasoiu commented on lib/controller.dart:
```diff
@@ -367,27 +389,40 @@ void command(UIAction action, Data data) {
       break;
     case UIAction.toggleActiveFilter:
       var d = data as ToggleActiveFilterData;
+      var filterDropdownID = generateFilterDropdownID(d.key);
+      var comparisonDropdownID = generateComparisonFilterDropdownID(d.key);
```
I don't think the controller should be doing this kind of translation for the view (ie. computing the ID of the view elements). If needed, the view can keep track of the mapping key -> DOM element, or you can just pass the key to the view, which can then compute the appropriate ID?

elayabharath replied:
That makes sense - have changed it to make the controller just pass the filter key and the view computes to toggle on/off the drop-downs. Also removed the previously used utils for generating the IDs.

marianamarasoiu reviewed (commented):
thanks @elayabharath - a first pass and a few comments

elayabharath committed:
Merge branch 'chart-config' into handle-user-interaction

elayabharath committed:
Show field values if labels are missing

elayabharath committed:
Merge branch 'chart-config' into handle-user-interaction

elayabharath committed:
Remove controller generating the view IDs

elayabharath committed:
Refactor to fold the filter id generators to view

elayabharath committed:
Apply incremental updates to filters than removing & adding all filters

elayabharath commented:
Thanks for the first pass @marianamarasoiu 

I have addressed the following changes
1. Removed the controller passing DOM IDs to the view for the filter elements
2. Changed the filters being thrown away and re-rendering. Now I enable / disable, hide / show the required filters
3. I'm still throwing away the charts and re-rendering. 

Could you take a look at this one more time and let me know? Thank you!

marianamarasoiu reviewed (approved):
looks great, thanks @elayabharath !

elayabharath commented:
Thank you @marianamarasoiu
